File: DP(Dynamic_Programming, 동적계획법)/백준_9461_파도반 수열/파도반 수열.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

long arr[101] = { 1,1,1,2,2 };

int main() {
	int T = 0;
	scanf("%d", &T);

	while (T--) {
		int n = 0;
		scanf("%d", &n);
		if (n < 5) printf("%lld\n", arr[n-1]);
		else {
			for (int i = 5; i < n; i++) {
				arr[i] = arr[i - 1] + arr[i - 5];
			}
			printf("%lld\n", arr[n - 1]);
		}
	}

	
}

File: DP(Dynamic_Programming, 동적계획법)/백준_9461_파도반 수열/파도반 수열.py
arr = [0 for _ in range(101)]
arr[1] = 1; arr[2] = 1; arr[3] = 1; arr[4] = 2; arr[5] = 2;
T = int(input())
for i in range(T):
    n = int(input())
    if n < 6: print(arr[n])
    else:
        for j in range(6, n+1):
            arr[j] = arr[j-1] + arr[j-5]
        print(arr[n])